### DISCUSSION TITLE
Plans to support conditional writes?

### DISCUSSION THREAD

**Discussion by purplefox** [minio/minio#20318] [Q&A] [answered]
As I'm sure you're all aware, S3 now supports conditional writes https://aws.amazon.com/about-aws/whats-new/2024/08/amazon-s3-conditional-writes/

This allows clients to maintain consensus directly in S3 without having to use something like etcd or zookeeper. 

Does Minio plan to support equivalent functionality, mimicking this new behaviour of the S3 API?

**Comment by harshavardhana** [comment, marked as answer]
We already do @purplefox we implemented it back in 2023 https://github.com/minio/minio/pull/16551 - AWS  simply copied our implementation. 

Of course, they overengineered it with DeleteObject() behavior we don't see a good need for that to be implemented yet. Until we see valid use-case.

**Reply by purplefox** [reply]
Thanks for the reply.

Can you confirm that your conditional write behaviour is the same as S3? According to the documentation for the new feature in S3 https://docs.aws.amazon.com/AmazonS3/latest/userguide/conditional-requests.html#conditional-writes

> If-None-Match — Upload the object only if no existing object with the same key name already exists in the specified bucket.

So it only puts the object if there's no object against the key, It does not check eTag.

But looking at your implementation, it seems to check eTag too, i.e. you only put the object if the eTag is different. This seems like different behaviour?

**Reply by harshavardhana** [reply]
@purplefox just use it and report any issues.

**Reply by purplefox** [reply]
If you claim S3 compatibility you don't think you should check it yourself?

**Reply by harshavardhana** [reply]
We check eTag as well our implementation is more complete, they implemented just `*`.  I.e if you know the right etag() you can avoid updating latest version not just check for existence. 

We also support `If-Match` not just `If-None-Match`. In this case AWS S3 is compatible with us not the other way around. @purplefox

**Reply by spacejam** [reply]
If I set IfNoneMatch=* with minio it doesn't prevent a second write to a particular key, as it does for s3. With minio I get two successful writes, instead of one success and then an expected 412 for the second write to the same key. Using minio `2023-08-09T23:30:22Z` and the rust `aws-sdk-s3` client version `1.50.0`. I've only been able to get minio to reject a second write if I'm able to specify the etag for an object that is already at that key as the value for IfNoneMatch, but this is not a semantic that I'm able to benefit from.

**Reply by klauspost** [reply]
@spacejam Upgrade.

**Reply by spacejam** [reply]
After upgrading to `2024-09-13T20-26-02Z` the behavior seems to match s3's as expected, thanks!

**Comment by purplefox** [comment]
I actually prefer the minio implementation, I wish S3 had implemented it
that way!

It's important that our product works exactly the same on MinIO as S3 so I
will check that IfNoneMatch=* works the same on both systems.

Thanks.

**Reply by purplefox** [reply]
One last thing - can I confirm that the Minio conditional update is also *atomic*. I.e. if two requests hit the key at the same time to do a conditional update, there is no chance of a race condition?

**Reply by harshavardhana** [reply]
That is the intention
